tpm2/test/nv_test.go

```diff
@@ -207,3 +208,228 @@ func TestNVAuthIncrement(t *testing.T) {
+	if err != nil {
+		t.Fatalf("Calling TPM2_NV_Read: %v", err)
+	}
+
+	// Verify the data is still the original data
+	expectedData := []byte{0x01, 0x02, 0x03, 0x04}
+	if !bytes.Equal(readRsp.Data.Buffer, expectedData) {
+		t.Errorf("Read data doesn't match expected data. Got %v, want %v", readRsp.Data.Buffer, expectedData)
+	}
+}
+
+func TestNVReadLock(t *testing.T) {
+	thetpm, err := simulator.OpenSimulator()
+	if err != nil {
+		t.Fatalf("could not connect to TPM simulator: %v", err)
+	}
+	defer thetpm.Close()
+
+	// Define the NV space with attributes that allow it to be locked for reading
+	def := NVDefineSpace{
+		AuthHandle: TPMRHOwner,
+		PublicInfo: New2B(
+			TPMSNVPublic{
+				NVIndex: TPMHandle(0x0180000F),
+				NameAlg: TPMAlgSHA256,
+				Attributes: TPMANV{
+					OwnerWrite:  true,
+					OwnerRead:   true,
+					ReadSTClear: true, // Allow TPM2_NV_ReadLock to lock this index
+				},
+				DataSize: 4,
+			}),
+	}
+	if _, err := def.Execute(thetpm); err != nil {
+		t.Fatalf("Calling TPM2_NV_DefineSpace: %v", err)
+	}
+
+	pub, err := def.PublicInfo.Contents()
+	if err != nil {
+		t.Fatalf("%v", err)
+	}
+	nvName, err := NVName(pub)
+	if err != nil {
+		t.Fatalf("Calculating name of NV index: %v", err)
+	}
+
+	// Write data to the NV space
+	write := NVWrite{
+		AuthHandle: AuthHandle{
+			Handle: TPMRHOwner,
+			Auth:   PasswordAuth(nil),
+		},
+		NVIndex: NamedHandle{
+			Handle: pub.NVIndex,
+			Name:   *nvName,
+		},
+		Data: TPM2BMaxNVBuffer{
+			Buffer: []byte{0x01, 0x02, 0x03, 0x04},
+		},
+		Offset: 0,
+	}
+	if _, err := write.Execute(thetpm); err != nil {
+		t.Errorf("Calling TPM2_NV_Write: %v", err)
+	}
+
+	// Read the data to verify it's accessible
+	read := NVRead{
+		AuthHandle: AuthHandle{
+			Handle: TPMRHOwner,
+			Auth:   PasswordAuth(nil),
+		},
+		NVIndex: NamedHandle{
+			Handle: pub.NVIndex,
+			Name:   *nvName,
+		},
+		Size:   4,
+		Offset: 0,
+	}
+	readRsp, err := read.Execute(thetpm)
+	if err != nil {
+		t.Fatalf("Calling TPM2_NV_Read before locking: %v", err)
+	}
+
+	// Verify the data is correct
+	expectedData := []byte{0x01, 0x02, 0x03, 0x04}
+	if !bytes.Equal(readRsp.Data.Buffer, expectedData) {
+		t.Errorf("Read data doesn't match expected data. Got %v, want %v", readRsp.Data.Buffer, expectedData)
+	}
+
+	// Lock the NV space against further reads
+	lock := NVReadLock{
+		AuthHandle: AuthHandle{
+			Handle: TPMRHOwner,
+			Auth:   PasswordAuth(nil),
+		},
+		NVIndex: NamedHandle{
+			Handle: pub.NVIndex,
+			Name:   *nvName,
+		},
+	}
+	if _, err := lock.Execute(thetpm); err != nil {
+		t.Errorf("Calling TPM2_NV_ReadLock: %v", err)
+	}
+
+	// Try to read from the NV space again, which should fail because it's locked
+	read2 := NVRead{
+		AuthHandle: AuthHandle{
+			Handle: TPMRHOwner,
+			Auth:   PasswordAuth(nil),
+		},
+		NVIndex: NamedHandle{
+			Handle: pub.NVIndex,
+			Name:   *nvName,
+		},
+		Size:   4,
+		Offset: 0,
+	}
+	if _, err := read2.Execute(thetpm); err == nil {
```

Comment:
oops, I forgot to mention the same suggestion as above here:

if _, err := read2.Execute(thetpm); !errors.Is(err, TPMRCNVLocked) {